### ISSUE TITLE
thread 'main' panicked at 'not currently active!?', src/tools/cargo/src/cargo/core/resolver/mod.rs:832:20

### ISSUE THREAD

**Issue by hw0lff** [rust-lang/cargo#11336]
### Problem

When building the linked minimal reproducible example cargo panics.

This bug got reported before in #11012 . However, they were not able to reproduce the error.
I found a [minimal reproducible example](https://gitlab.com/w0lff/cargo-issue-11012).

### Steps

1. `git clone https://gitlab.com/w0lff/cargo-issue-11012.git`
2. `cargo build`

### Notes

The `=` in the version field of the `wayland-client` crate is needed now because the maintainer published newer versions. Without it cargo fails with an error and does not panic anymore.

### Version

```text
cargo 1.65.0 (4bc8f24d3 2022-10-20)
release: 1.65.0
commit-hash: 4bc8f24d3e899462e43621aab981f6383a370365
commit-date: 2022-10-20
host: x86_64-unknown-linux-gnu
libgit2: 1.5.0 (sys:0.15.0 vendored)
libcurl: 7.83.1-DEV (sys:0.4.55+curl-7.83.1 vendored ssl:OpenSSL/1.1.1q)
os: Arch Linux Rolling Release [64-bit]
```

**Comment by weihanglo** [comment]
Thanks for your report!

It is definitely a bug that Cargo panics. From the other angle, it also does things partially correctly. Your package depends on two SemVer-compatible versions of a single package, which is not currently possible in Cargo.

I've got a minimal example in Cargo's own test dialect more or less the same as your example. Let's investigate it more!

<details>

<summary>Expand to see the example</summary>

```rust
#[cargo_test]
fn mismatched_version2() {
    Package::new("qux", "0.1.0-beta.1").publish();
    Package::new("qux", "0.1.0-beta.2").publish();
    Package::new("bar", "0.1.0").dep("qux", "=0.1.0-beta.1").publish();
    let p = project()
        .file(
            "Cargo.toml",
            r#"
                 [package]
                 name = "foo"
                 version = "0.1.0"

                 [dependencies]
                 bar = "0.1.0"
                 qux = "=0.1.0-beta.2" # still panic without `=`

                 [patch.crates-io]
                 qux = { path = "qux" }
            "#,
        )
        .file("src/lib.rs", "")
        .file(
            "qux/Cargo.toml",
            r#"
                [package]
                name = "qux"
                version = "0.1.0-beta.1"
            "#,
        )
        .file("qux/src/lib.rs", "")
        .build();

    p.cargo("check")
        .with_status(101)
        .with_stderr("")
        .run();
}
```

</details>

**Comment by tomasro27** [comment]
I am seeing the same issue as well with a combination of crates, but it is hard to debug given that cargo panics and does not update lock file.

Any updates on fixing the panic?

**Comment by weihanglo** [comment]
Fixed by #11770 in 1.70.0. Thanks for the report and reproducer!
